File: python/394_decode_string.py
class Solution(object):
    def decodeString(self, s):

        idx = 0
        stack = [[0, '']] # multiplier, string

        while idx < len(s):

            if s[idx].isalpha():
                stack[-1][1] += s[idx]
                idx += 1

            elif s[idx].isdigit():
                left = s.find('[', idx)
                mult = int(s[idx:left])
                idx = left + 1
                stack.append([mult, ''])

            else:
                stack[-2][1] += stack[-1][0] * stack[-1][1]
                stack.pop()
                idx += 1

        return stack[0][1]

File: cpp/394_decode_string.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    string decodeString(string s) {

		stack<pair<string, int>> stack;
		stack.push(make_pair("", 1));

		for (int i = 0; i < s.length(); i++) {
			if (isdigit(s[i])) {
				int left_bracket_index = s.find('[', i);
				int times = stoi(s.substr(i, left_bracket_index - i));
				stack.push(make_pair("", times));
				i = left_bracket_index;
			}
			else if (s[i] == ']') {
				auto top = stack.top();
				stack.pop();
				for (int j = 0; j < top.second; j++) {
					stack.top().first += top.first;
				}
			}
			else {
				stack.top().first += s[i];
			}
		}

		return stack.top().first;
    }
};
